overlord/snapstate/handlers.go

```diff
@@ -1506,6 +1511,12 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
 		linkCtx.RequireMountedSnapdSnap = true
 	}
+
+	// write sequence file for failover helpers
+	if err := writeSeqFile(snapsup.InstanceName(), snapst); err != nil {
```

Comment:
looking at how this is used by cmd/snap-failure this change looks fine, even an improvement but it would be good if other people familiar with snap-failure like @anonymouse64  and @bboozzoo double check the reasoning here

Reply:
I think this change is good, since at the start of doLinkSnap, we will set in memory snapst.Current to this revision we are trying to link, which means that if snapd dies as part of this change/refresh and snap-failure comes up, then snap-failure will see that the `Current` setting was the revision being linked, and it will go back to the previous one. I think actually as you say it is slightly better than the status quo, as it seems to be there exists a possibility during this function where we haven't committed the new `Current` setting to the sequence file, so if we die and snap-failure comes up, it will think not the previous revision, but the previous previous revision is the one it should revert back to. Though I could be missing some mechanism by which we avoid that actually.

I actually don't wonder if we shouldn't write the sequence file immediately when we set snapst.Current to this one? Though it's unclear how snapd would die between the current place we set it in memory and when we write it to the sequence file, so it's probably fine.

Reply:
If we have revisions a and b, and installing rev c, we write it here, we have `seq: [a,b,c], current: c`. Should snapd fail, snap-failure will start, pick up revision preceding 'c', so it's 'b' and it will run snapd explicitly through `/snap/snap/b/usr/lib/snapd`. Before it was IMO more complicated, as when snapd rev `b` died before having written the seq file (which happened after link-snap), snap-failure would run revision `a` (as we have `seq: [a, b], current b`), wouldn't it? Anyway, AFAIU snapd would keep on executing in snap-failure now.

The only time snap-failure looks a the current symlink is when there is 'previous' revision, but it already implies that there is just one revision of snapd.